backend/src/write-side/inviting-applicants/domain/first-name.value-object.ts

```diff
@@ -0,0 +1,12 @@
+export class FirstName {
```

Comment:
Może te wszystkie najprostsze ValueObject-y da się uprościć, co wy na to?
```javascript
function ValueObject<N, T>(policy: (_: T) => void = (x: T) => {}) {
    return class _ {
        constructor(readonly raw: T) {};

        static from(val: T) {
            policy(val);
            return new _(val);
        }
    };
};

const EmptyStringPolicy = (name: string) => {
            if (name.length <= 0) {
                throw new Error('Name cannot be empty!');
            }};

const EmailPolicy = (email: string) => {
    if (!email.includes('@')) {
      throw new Error('Invalid email!');
    }};


namespace ValeObjects {
    export const LastName = ValueObject<"LastName", string>(EmptyStringPolicy);
    export const PersonalEmail = ValueObject<"PersonalEmail", string>(EmailPolicy);
}

const x = ValeObjects.LastName.from("ala");
console.log(x.raw);
const y = ValeObjects.PersonalEmail.from("ala");
```
[Playground Link](https://www.typescriptlang.org/play?ts=3.6.3#code/GYVwdgxgLglg9mABANQIYBsQFMDyAjAKy2gB4A5AGkQBUA+ACgAc50YIBPALkXoH1vqASkQBeWogBucGABNRPAB4DhYxAG8AvsLUAoRPsQAnLFBCGkEdKgDO1xL3V6DziAmtRDIaHEP1jqGQR0diNUAHdldQ0Abh0nZ313VFgIRGBDOABbegkMSN0EwsRmVg4cjEFYooTjU3NEMCww+3L0SviEjQ6YnR6dVzB3RABRTMYodgBlDxgwAHMABRY2EJEeMFRMrG53Q1m5lXEC6oMYYHXNrAA6dCx5qAALRBI1gAZtDpODR4zmxubhoYMr4AORkS6ICCoMBgOBQRB4LCILBjCYAQhB7S++g0fQGQ1GqBg6CWpVWPBRRPQOxm80OjmcZx4aMpxKus0sIBkWGs9BBAAFMR9Cj84H8miMgT4+QBJMC5VhyVnoDFYgy42JxDZbayMVAQJFoW74IjQOzHfRYBTMQzw-HwgAyNig4K28iN2BNxCgJAARE73K6sL6qLt9gxRuMprTFssOGrLdafHa3PCFlhDNYEBhCcT3RhPYRvX705ns+hc+gQ4gw-MI5kqaSVu0uv1U4gFPnjUWzVcAy7Lld0ll6L6MKhfe18Sxrug4HN6AoroZwlP2+Sjbge1BrFdS1mNhWG2zh9kx1ZJ7EgA)

Reply:
Jestem jak najbardziej za wprowadzeniem tutaj jakiejś wspólnej abstrakcji. Nad tym muszę się chwilę pochylić :) Nie wiem czy by spełniały wszystkie zadania. Obczaj sobie jeszcze to: 
https://medium.com/@hannespetri/type-safe-value-objects-in-typescript-d1b119c4f5cd

Ale kurde, wygląda mi to spoko co pokazałeś :) Nawet do tych policy możnaby wprowadzać array (albo kompozycję), wtedy jakieś w stylu EmptyStringPolicy mogłyby być reużywalne.

Tylko nie działa mi w tym przykładzie jedna rzecz: 
ValueObject powinien działać tak, że jest sobie równe:
``` ValeObjects.LastName.from("ala") ===  ValeObjects.LastName.from("ala")```
Ale już np poniższy przykład zwróci false: 
``` ValeObjects.FirstName.from("ala") ===  ValeObjects.LastName.from("ala")```
Czyli, żeby te typy były rozróżnialne.

Reply:
To już o za dużo prosisz xD
W TS/JS nie można przeciążać operatorów więc nie da się zamienić funkcjonalności '===' który porównuje dla obiektów referencję. A jak byś nie starał nie da się stworzyć własnego primitive data type.

Reply:
Jedyne co można zrobić to implementacja metod/funkcji typu `eq`, `ge`, `le` itd., ewentualnie [valueOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf) ale ono też nie działa super i może powodować mindfuck co ten JS wyprawia :D

Reply:
Zostawiłbym to na później :) Żeby nie blokować dalszej pracy i rozsrastać PR. A obczaje możliwości.